### PR TITLE
Computation of the WcsMap kernel at the nearest valid exposure

### PR DESCRIPTION
**Pull request by bkhelifi**
**Description**

This pull request fixes the bug mentioned in the Issue #3707. It changes the computation of the kernel, previously done with the exposure at the centre of the map and now with the mean non-null exposure over the whole map.

**Dear reviewer**
In addition to the change of the computation logics, I have added a test on the kernel values and also a test with a map with bins with null exposure.

### CODE REVIEW

File: gammapy/estimators/map/ts.py [hunk 1]
```diff
@@ -236,10 +236,9 @@ def estimate_kernel(self, dataset):
         model = self.model.copy()
         model.spatial_model.position = geom.center_skydir
 
-        # Creating exposure map with exposure at map center
+        # Creating exposure map with the mean non-null exposure
         exposure = Map.from_geom(geom, unit="cm2 s1")
-        exposure_center = dataset.exposure.to_region_nd_map(geom.center_skydir)
-        exposure.data[...] = exposure_center.data
+        exposure.data[...] = np.mean(dataset.exposure.data[dataset.exposure.data > 0.])
```
Comment (QRemy) [inline gammapy/estimators/map/ts.py:241]: you could also consider the dataset.mask and only finite values to compute the mean exposure :
```python
exposure_mask = (dataset.mask.data) & (dataset.exposure.data > 0.) & np.isfinite(dataset.exposure.data)
exposure.data[...] = np.mean(dataset.exposure.data[exposure_mask])
```

Comment (QRemy) [reply gammapy/estimators/map/ts.py:241]: this also address my comment about unit

Comment (QRemy) [reply gammapy/estimators/map/ts.py]: I think we can use  dataset.mask.interp_to_geom( dataset.exposure.geom)  as a the mask in reco energy is dilated compared to the mask in true energy we don't loose information .

Comment (bkhelifi) [reply gammapy/estimators/map/ts.py]: This can not work as the energy axis names are different between the exposure (in true energy) and the mask (in reco energy)... No turn-around found yet.

**Review by registerrier — COMMENTED** [review]
Thanks @bkhelifi .

I am not sure you need to use a full mask on the exposure, I suppose that the image_mask should be sufficient. Note that the current implementation does not return an array for the exposure.
Is it possible to quantify the result when applying the mean compared to the center value? 

Is it possible that no PSF is available as well?

**Comment by adonath** [comment]
Idea from dev meeting today: make behavior consistent with IRF kernel extraction. So find the nearest valid position to extract the exposure from. I think for a TS map we always need a PSF, so we could use:

```python3
datasets.psf._get_nearest_valid_position(position=, )
```
It's not ideal to rely on a hidden method, but as we do it internally in Gammapy it's acceptable.

File: gammapy/estimators/map/ts.py [hunk 2]
```diff
@@ -236,10 +236,9 @@ def estimate_kernel(self, dataset):
         model = self.model.copy()
         model.spatial_model.position = geom.center_skydir
 
-        # Creating exposure map with exposure at map center
+        # Creating exposure map with the mean non-null exposure
         exposure = Map.from_geom(geom, unit="cm2 s1")
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/estimators/map/ts.py:241]: This is still unaddressed?

File: gammapy/datasets/utils.py [hunk 3]
```diff
@@ -23,3 +23,12 @@ def get_axes(ax1, ax2, width, height, args1, args2, kwargs1=None, kwargs2=None):
         raise ValueError("Either both or no Axes must be provided")
 
     return ax1, ax2
+
+
+def get_nearest_valid_exposure_position(exposure, position=None):
+    import numpy as np
```
Comment (registerrier) [inline gammapy/datasets/utils.py:29]: I would import globally here.

File: gammapy/estimators/map/ts.py [hunk 4]
```diff
@@ -236,10 +237,13 @@ def estimate_kernel(self, dataset):
...
-        exposure_center = dataset.exposure.to_region_nd_map(geom.center_skydir)
-        exposure.data[...] = exposure_center.data
+        position = get_nearest_valid_exposure_position(dataset.exposure, geom.center_skydir)
+        exposure_position = dataset.exposure.to_region_nd_map(position)
+        if not np.any(exposure_position.data):
+            raise ValueError("The exposure is null. Impossible to convolve the model with the PSFKernel.")
```
Comment (registerrier) [inline gammapy/estimators/map/ts.py:245]: I would use a simpler error message e.g.:
`No valid exposure. Impossible to compute kernel for TS Map.`

File: gammapy/estimators/map/ts.py [hunk 5]
```diff
@@ -215,7 +216,7 @@ def selection_all(self):
     def estimate_kernel(self, dataset):
         """Get the convolution kernel for the input dataset.
 
-        Convolves the model with the PSFKernel at the center of the dataset.
+        Convolves the model with the IRFs.
```
Comment (registerrier) [inline gammapy/estimators/map/ts.py:219]: Maybe specify: `with the IRFs at the center of the dataset (or the nearest position with non-null exposure)`.

File: gammapy/estimators/map/ts.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (registerrier) [reply gammapy/estimators/map/ts.py:241]: Note that this is a pre-existing code, and possibly buggy indeed because what is passed is the `exposure.data`. 
A possible solution is to replace `"cm2s"` by `dataset.exposure.unit`

**Review by registerrier — CHANGES_REQUESTED** [review]
Thanks @bkhelifi . This looks good. 
I have left a few minor comments. 
Please remove the `assert_raises` which we never use in gammapy and use instead `with pytest.raises`.

File: gammapy/estimators/map/ts.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkhelifi) [reply gammapy/estimators/map/ts.py:241]: But this contradicts a bit the initial comment of Quentin... What's happening if `data.exposure.unit` is not in `"cm2s"`?
This code allows to fix this case...
What is the consensus then?
